### PR TITLE
Onboarding - Enable biometric unlock

### LINKED ISSUE #30: Enable biometric unlock (onboarding)
## Acceptance criteria
* Should provide a way to skip this step if user chooses not to enable
* On choosing to enable, user should be prompted to authenticate with biometrics
* Only display this to users that: 
  1. Have Fingerprint available on their device  
  2. Have this feature enabled on their device (have a Fingerprint set).  

Make sure that navigating to this screen prevents the user from seeing the ugly webpage error as referenced in #273

requires #28 

### Engineering Notes

On iOS, this required "pausing" in driving the UI from the datastore state while onboarding. We added a `val onboarding: Observable<Boolean>` to the `RouteStore`, and ignored incoming non-onboarding routes while it was true.

### Visual Design

**InVision**
Initial: https://mozilla.invisionapp.com/share/3SN35UHWJDX#/309979870_Fingerprint
Error: https://mozilla.invisionapp.com/share/3SN35UHWJDX#/310610373_Fingerprint_-_Error
Success*: https://mozilla.invisionapp.com/share/3SN35UHWJDX#/311048434_Fingerprint_-_Success
*Intent would be that the success message displays, pauses for a second, then transitions to next screen in onboarding (AutoFIll)

**Zeplin**
Initial: https://zpl.io/VYBxqRk
Error: https://zpl.io/Vk5l1nW
Succes: https://zpl.io/25qK4rW

### PR DESCRIPTION
**Pull request by eliserichards**
Fixes #30

## Testing and Review Notes
@nickbrandt noticed a crash/failure happening when the fingerprint fails five times on onboarding:
> I did notice an issue where if you have 5 failed attempts, and it transitions automatically on to the entry list, ... If I lock the device and go to unlock,I’m presented with the Fingerprint dialog. This also caused a crash on my device, but then I was able to unlock using our Fingerprint dialog.

This seems to be fixed from my end (I can't reproduce it) so it might already have been fixed during another refactoring, but we may want to open a bug just to track it in case we encounter the crash again. 

## Screenshots or Videos
**Initial onboarding screen**
image

**Successful fingerprint authentication**
image

**Failed/errored fingerprint authentication**
image
### eng todo
- [x] display iff fingerprint is available on that device
- [x] display if fingerprint unlock is already enabled on to unlock their device
- [x] when "touch sensor" -> begin the fingerprint auth flow
- [x] skip for now continues onto itemlist route
- [x] if "touch sensor" succeeds -> 
  - [x] show success design 
  - [x] and enable in settings
- [x] if "touch sensor" fails -> show failure design 
  ~~- questions: should we prompt to retry on onboarding page? does fingerprint auth flow allow retry? if this onboarding step fails do we automatically transition? - **no**~~

### review todo
- [x] add “WIP” to the PR title if pushing up but not complete nor ready for review
- [x] double check the original issue to confirm it is fully satisfied
- [x] add testing notes and screenshots in PR description to help guide reviewers
- [x] add unit tests
  - optional: consider adding instrumentation (integration/UI) tests
- consider running this branch in a debug simulator and check for memory leak notifications or any warnings
- [x] request the "UX" team perform a design review (if/when applicable)
- [x] make sure CI builds are passing (e.g.: fix lint and other errors)
- [x] check on the [accessibility](https://mozilla-lockbox.github.io/lockbox-android/accessibility/) of any added UI

### CODE REVIEW

File: app/src/main/java/mozilla/lockbox/store/RouteStore.kt [hunk 1]
```diff
@@ -36,14 +45,40 @@ class RouteStore(
...
+        // do we want to onboard if they disconnect their account and
+        // then re-login?
+        dataStore.state
+            .subscribe { state ->
+                when (state) {
+                    is DataStore.State.Unprepared -> onboardingState.onNext(true)
+                    else -> onboardingState.onNext(false)
+                }
+            }
+            .addTo(compositeDisposable)
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/store/RouteStore.kt]: See how this is done above:

```kotlin
dataStore.state
    .map { it == DataStore.State.Unprepared }
    .subscribe(onboardingState)
```

Then, no need for a `compositeDisposable`.

File: app/src/main/java/mozilla/lockbox/presenter/OnboardingFingerprintAuthPresenter.kt [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+            // OnboardingAction.OnDismiss can be used in the next onboarding screens too
+            view.onDismiss
+                .subscribe {
+                    dispatcher.dispatch(OnboardingAction.OnDismiss)
+                }.addTo(compositeDisposable)
+        } else {
+            // just skip the route altogether, not follow Dismiss logic
+            dispatcher.dispatch(RouteAction.SkipOnboarding)
+        }
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/presenter/OnboardingFingerprintAuthPresenter.kt]: The whole of this `if/else` statement should be in the `subscribe` block in L44. 

You may want to consider: `routeStore.onboarding.take(1)`.

File: app/src/main/java/mozilla/lockbox/store/RouteStore.kt [hunk 3]
```diff
@@ -36,14 +45,40 @@ class RouteStore(
...
+            // bad. figure out how to make this work again
+            else -> null!!
+        }
+    }
+
+    private var triggerOnboarding: Boolean? = null
+    private val triggerOnboardingSubject: Consumer<Boolean>
+        get() = Consumer { onboarding ->
+            triggerOnboarding = onboarding
+        }
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/store/RouteStore.kt]: We should be avoiding having `Consumers` in stores. Stores should be listening to `dispatcher` for most of the time, and occasionally other stores.

AFAIUI, this should be a `Subject<Boolean>`.

File: app/src/main/java/mozilla/lockbox/view/OnboardingAutofillFragment.kt [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import mozilla.lockbox.R
+
+class OnboardingAutofillFragment : DialogFragment() {
+
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+        return inflater.inflate(R.layout.fragment_onboarding, container, false)
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/view/OnboardingAutofillFragment.kt]: Should this have a presenter associated with it?

**Comment by eliserichards** [comment]
@jhugman @ioana-farcas @linuxwolf I don't think I understand enough about the fingerprint authentication, so please bear with me if this is a stupid question: 

Does a user need to have fingerprint security enabled BEFORE they use the app in order to prompt the fingerprint dialog? I am seeing this being used, specifically the `hasEnrolledFingerprints()`: 
```open val isFingerprintAuthAvailable: Boolean
  get() = fingerprintManager.isHardwareDetected && fingerprintManager.hasEnrolledFingerprints()
```

In my implementation I would assume that, even if a user has no fingerprint set up, they would still be prompted to set up security on their device with the onboarding screen. Is this a bad assumption, and should I mimic the logic that we already have for the fingerprint dialog and use `isFingerprintAuthAvailable`?

**Comment by eliserichards** [comment]
@nickbrandt - From what I can find, the "5 tries = too many attempts" is an Android FingerprintManager functionality. I am using the same logic that we have in the fingerprint auth dialog. I'm not sure if it would be a good idea to change that right now, and I would like to ask Sasha & James for their opinions if that's ok with you. 

Re: disconnecting your account not triggering onboarding -- I am seeing this functionality working on my device 🤔 I can try downloading the build and checking if it's working the same there.

**Comment by nickbrandt** [comment]
@eliserichards Ah gotcha, didn’t realize closing out of the screen was part of that logic. I did notice an issue where if you have 5 failed attempts, and it transitions automatically on to the entry list, when I go to settings Fingerprint is shown as enabled. If I lock the device and go to unlock,I’m presented with the Fingerprint dialog. This also caused a crash on my device, but then I was able to unlock using our Fingerprint dialog. 

Apologies about the disconnect issue, I must be losing it as I’m no longer experiencing that problem (disconnecting causes onboarding to occur on next sign in)

File: app/src/main/java/mozilla/lockbox/action/SettingAction.kt [hunk 5]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/action/SettingAction.kt]: 👾 

this kind of clutter can make PRs hard to read!

Comment (sashei) [inline app/src/main/java/mozilla/lockbox/action/SettingAction.kt]: 👾 

this kind of clutter can make PRs hard to read!

File: app/src/androidTest/java/mozilla/lockbox/robots/OnboardingFingerprintRobot.kt [hunk 6]
```diff
@@ -0,0 +1,16 @@
+package mozilla.lockbox.robots
```
Comment (sashei) [inline app/src/androidTest/java/mozilla/lockbox/robots/OnboardingFingerprintRobot.kt]: 🤖 🎉 awesome!

File: app/src/main/java/mozilla/lockbox/adapter/SortItemAdapter.kt [hunk 7]
```diff
@@ -26,7 +26,7 @@ class SortItemAdapter(
         val label = super.getView(position, convertView, parent) as TextView
         label.setTextAppearance(R.style.TextAppearanceWidgetEventToolbarTitle)
         label.setTextColor(Color.WHITE)
-        label.setBackgroundColor(label.resources.getColor(R.color.colorPrimary, null))
+        label.setBackgroundColor(label.resources.getColor(R.color.color_primary, null))
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/adapter/SortItemAdapter.kt]: 🐫 ➡️ 🐍 

👍 👍 👍

File: app/src/main/java/mozilla/lockbox/presenter/FxALoginPresenter.kt [hunk 8]
```diff
@@ -23,8 +23,8 @@ import mozilla.lockbox.support.Constant
 interface FxALoginView {
     var webViewRedirect: ((url: Uri?) -> Boolean)
     val skipFxAClicks: Observable<Unit>?
-    fun handleNetworkError(networkErrorVisibility: Boolean)
     val retryNetworkConnectionClicks: Observable<Unit>
+    fun handleNetworkError(networkErrorVisibility: Boolean)
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/presenter/FxALoginPresenter.kt]: is this related to this PR?

File: app/src/main/java/mozilla/lockbox/presenter/OnboardingFingerprintAuthPresenter.kt [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+@ExperimentalCoroutinesApi
+class OnboardingFingerprintAuthPresenter(
+    private val view: OnboardingFingerprintView,
+    private val dispatcher: Dispatcher = Dispatcher.shared,
+    private val fingerprintStore: FingerprintStore = FingerprintStore.shared
+) : Presenter() {
+
+    override fun onViewReady() {
+        if (fingerprintStore.isFingerprintAuthAvailable) {
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/presenter/OnboardingFingerprintAuthPresenter.kt]: tidiness nit: rather than having the bulk of the method's logic inside of an `if` statement, consider:
```kotlin
if (!fingerprintStore.isFingerprintAuthAvailable) {
    dispatcher.dispatch(RouteAction.SkipOnboarding)
    return
}
// rest of the code from inside `if`
```

File: app/src/main/java/mozilla/lockbox/store/RouteStore.kt (hunk 1 above)
Comment (sashei) [reply app/src/main/java/mozilla/lockbox/store/RouteStore.kt]: I'm resurrecting James' comment here; this has not been resolved yet

File: app/src/main/java/mozilla/lockbox/view/FxALoginFragment.kt [hunk 10]
```diff
@@ -83,4 +82,9 @@ class FxALoginFragment : BackableFragment(), FxALoginView {
 
     override val skipFxAClicks: Observable<Unit>?
         get() = skipFxA?.clicks()
+
+    override fun onDestroyView() {
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/view/FxALoginFragment.kt]: This is already in the `Fragment.kt` class and is unnecessary here.

File: app/src/main/java/mozilla/lockbox/view/FxALoginFragment.kt [hunk 11]
```diff
@@ -30,7 +30,6 @@ import mozilla.lockbox.support.isDebug
 
 @ExperimentalCoroutinesApi
 class FxALoginFragment : BackableFragment(), FxALoginView {
-
     private var errorHelper = NetworkErrorHelper()
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/view/FxALoginFragment.kt]: 👾

File: app/src/main/java/mozilla/lockbox/view/OnboardingFingerprintAuthFragment.kt [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+
+    private val _authCallback = PublishSubject.create<AuthCallback>()
+    override val authCallback: Observable<AuthCallback> get() = _authCallback
+
+    override val onDismiss: Observable<Unit>
+        get() = view!!.skipOnboarding.clicks()
+
+    private var isEnablingDismissed: Boolean = true
+
+    sealed class AuthCallback : FingerprintAuthCallback {
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/view/OnboardingFingerprintAuthFragment.kt]: why did we keep an `AuthCallback` around, but rename the root class to `FingerprintAuthCallback`? This seems like some unnecessary shuffling to me.

File: app/src/main/java/mozilla/lockbox/presenter/SettingPresenter.kt [hunk 13]
```diff
@@ -74,16 +74,20 @@ class SettingPresenter(
             .subscribe {
                 if (it is FingerprintAuthAction.OnAuthentication) {
                     when (it.authCallback) {
-                        is FingerprintAuthDialogFragment.AuthCallback.OnAuth -> dispatcher.dispatch(
-                            SettingAction.UnlockWithFingerprint(true)
-                        )
+                        is FingerprintAuthDialogFragment.AuthCallback.OnAuth ->
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/presenter/SettingPresenter.kt]: please avoid committing formatting changes without any actual code changes! 🙇‍♀️

File: app/src/main/java/mozilla/lockbox/store/RouteStore.kt [hunk 14]
```diff
@@ -19,8 +23,15 @@ import mozilla.lockbox.support.asOptional
 @ExperimentalCoroutinesApi
 class RouteStore(
     dispatcher: Dispatcher = Dispatcher.shared,
-    dataStore: DataStore = DataStore.shared
+    dataStore: DataStore = DataStore.shared,
+    fingerprintStore: FingerprintStore = FingerprintStore.shared
 ) {
+    internal val compositeDisposable = CompositeDisposable()
+
+    private val onboardingState: ReplaySubject<Boolean> = ReplaySubject.createWithSize(1)
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/store/RouteStore.kt]: a few things:

let's put all the properties _after_ the `companion object` declaration so that we don't have the lonely `routes` object down below, and unify our `Observable` declarations so that they all look like:

```kotlin
val onboardingState: Observable<Boolean> = ReplaySubject.createWithSize(1)
```

File: app/src/main/java/mozilla/lockbox/store/RouteStore.kt [hunk 15]
```diff
@@ -32,18 +43,48 @@ class RouteStore(
             .filterByType(RouteAction::class.java)
             .subscribe(routes as Subject)
 
+        this.onboarding
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/store/RouteStore.kt]: phew! this is all pretty hairy; I'd love to take some time this afternoon to talk to you about the routing approach here.

File: app/src/main/java/mozilla/lockbox/presenter/FxALoginPresenter.kt (hunk 8 above)
Comment (eliserichards) [reply app/src/main/java/mozilla/lockbox/presenter/FxALoginPresenter.kt]: no, looks like it just got moved around

File: app/src/main/java/mozilla/lockbox/presenter/OnboardingFingerprintAuthPresenter.kt (hunk 9 above)
Comment (eliserichards) [reply app/src/main/java/mozilla/lockbox/presenter/OnboardingFingerprintAuthPresenter.kt]: 👏

File: app/src/main/java/mozilla/lockbox/store/RouteStore.kt (hunk 14 above)
Comment (eliserichards) [reply app/src/main/java/mozilla/lockbox/store/RouteStore.kt]: 👏

File: app/src/main/java/mozilla/lockbox/view/FxALoginFragment.kt (hunk 10 above)
Comment (eliserichards) [reply app/src/main/java/mozilla/lockbox/view/FxALoginFragment.kt]: 👏

File: app/src/main/java/mozilla/lockbox/presenter/FingerprintDialogPresenter.kt [hunk 16]
```diff
@@ -29,8 +29,7 @@ class FingerprintDialogPresenter(
     private val view: FingerprintDialogView,
     private val dispatcher: Dispatcher = Dispatcher.shared,
     private val fingerprintStore: FingerprintStore = FingerprintStore.shared
-) :
-    Presenter() {
+) : Presenter() {
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/presenter/FingerprintDialogPresenter.kt]: nit: unrelated whitespace change

File: app/src/main/java/mozilla/lockbox/presenter/SettingPresenter.kt (hunk 13 above)
Comment (sashei) [reply app/src/main/java/mozilla/lockbox/presenter/SettingPresenter.kt]: change still not made :(

File: app/src/main/java/mozilla/lockbox/store/RouteStore.kt [hunk 17]
```diff
@@ -25,15 +33,29 @@ class RouteStore(
...
+    internal val compositeDisposable = CompositeDisposable()
+
+    private val onboarding: Observable<Boolean> = BehaviorRelay.createDefault(false)
+
     val routes: Observable<RouteAction> = ReplaySubject.createWithSize(1)
 
     init {
         dispatcher.register
             .filterByType(RouteAction::class.java)
+            .debug("dispatched route actions")
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/store/RouteStore.kt]: looks like you committed a few extra `.debug`s... generally, these aren't good to leave in production code!!

File: app/src/main/java/mozilla/lockbox/model/FingerprintAuthCallback.kt [hunk 18]
```diff
@@ -0,0 +1,12 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package mozilla.lockbox.model
+
+interface FingerprintAuthCallback {
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/model/FingerprintAuthCallback.kt]: can we remove the `AuthCallback` subclasses and just use this, please?

File: app/src/main/java/mozilla/lockbox/presenter/FingerprintDialogPresenter.kt (hunk 16 above)
Comment (sashei) [reply app/src/main/java/mozilla/lockbox/presenter/FingerprintDialogPresenter.kt]: ...on re-review, this is a good improvement :p fine for now.

File: app/src/main/java/mozilla/lockbox/store/RouteStore.kt (hunk 17 above)
Comment (eliserichards) [reply app/src/main/java/mozilla/lockbox/store/RouteStore.kt]: shoot! good call
